src/notebookConcatDocument.ts

```diff
@@ -409,106 +579,235 @@ export class NotebookConcatDocument implements vscode.TextDocument, vscode.Dispo
+        const concatSpan = this._spans.find((r) => concatOffset >= r.startOffset && concatOffset < r.endOffset);
+        if (concatSpan) {
+            // Diff is into the concat span
+            const diff = concatOffset - concatSpan.startOffset;
 
-        for (let i = 0; i < offset; i += 1) {
-            if (this._contents[i] === '\n') {
-                lineCounter += 1;
+            // If real cell, then just add real offset
+            if (concatSpan.inRealCell) {
+                return diff + concatSpan.realOffset;
             }
+
+            // If not a real cell, just use the plain real offset.
+            return concatSpan.realOffset;
         }
+        return concatOffset;
+    }
 
-        return lineCounter;
-    }
-
-    private changeRange(
-        newText: string,
-        from: vscode.Position,
-        to: vscode.Position,
-        cellIndex: number
-    ): protocol.TextDocumentContentChangeEvent[] {
-        const fromOffset = this.convertToOffset(from);
-        const toOffset = this.convertToOffset(to);
-
-        // Recreate our contents, and then recompute all of our lines
-        const before = this._contents.substring(0, fromOffset);
-        const after = this._contents.substring(toOffset);
-        this._contents = `${before}${newText}${after}`;
-        this._lines = this.createLines();
-
-        // Update ranges after this. All should move by the diff in length, although the current one
-        // should stay at the same start point.
-        const lengthDiff = newText.length - (toOffset - fromOffset);
-        for (let i = cellIndex; i < this._cellRanges.length; i += 1) {
-            if (i !== cellIndex) {
-                this._cellRanges[i].startOffset += lengthDiff;
-            }
-            this._cellRanges[i].endOffset += lengthDiff;
+    private getEditPosition(cellUri: vscode.Uri, offset: number): vscode.Position {
+        // Get all the lines for this cell
+        const cellLines = this._lines.filter((l) => l.cellUri.toString() == cellUri.toString());
+
+        // Find the line that contains this offset. If not found, use the last line in the cell
+        const line = cellLines.find((l) => offset >= l.offset && offset < l.endOffset);
+        if (line) {
+            return new vscode.Position(line.lineNumber, offset - line.offset);
+        } else {
+            // Otherwise we're off the end of the cell
+            return new vscode.Position(
+                cellLines[cellLines.length - 1].lineNumber,
+                cellLines[cellLines.length - 1].text.length
+            );
         }
+    }
+
+    private createSpan(
+        cellUri: vscode.Uri,
+        text: string,
+        realText: string,
+        offset: number,
+        realOffset: number
+    ): NotebookSpan {
+        // Compute fragment based on cell uri
+        const fragment =
+            cellUri.scheme === InteractiveInputScheme ? -1 : parseInt(cellUri.fragment.substring(2) || '0');
```

Comment:
Doesn't the fallback need to be 0 not '0'?

Reply:
No actually it's inside the parse. So it has to parse the '0'

Reply:
Whoops totally missed that. It just looked odd, I see it now.